apollo-federation/src/correctness/response_shape_compare.rs

```diff
@@ -254,56 +275,215 @@ fn compare_possible_definitions_per_type_condition<T: PathConstraint>(
+                     this_clause = this_variant.boolean_clause()
+                )));
+            };
+            let Some(other_variant) = collect_variants_for_boolean_condition(other, &full_clause)? else {
+                return Err(ComparisonError::new(format!(
+                    "no variants found for Boolean condition in solve_boolean_constraints: {full_clause}"
+                )));
+            };
+            compare_definition_variant(path_constraint, &full_clause, this_variant, &other_variant)
+                .map_err(|e| {
+                    e.add_description(&format!(
+                        "mismatched variants for hypothesis: {hypothesis}\n\
+                         - Assumption: {assumption}\n\
+                         - this_clause: {this_clause}\n\
+                         - Full condition: {full_clause}",
+                        this_clause = this_variant.boolean_clause())
+                    )
+                })
+        });
+        match result {
+            Ok(()) => {
+                return Ok(());
+            }
+            Err(e) => {
+                let group_str = group.iter().join(", ");
+                errors.push(format!(
+                    "solve_boolean_constraints: group: {group_str}\n\
+                    detail: {e}",
+                ));
+            }
+        }
+    }
+    // None worked => error
+    Err(ComparisonError::new(format!(
+        "Failed to solve Boolean constraints w/ assumption {assumption}\n\
+         this_variant: {this_variant}\n\
+         other: {other}\n\
+         detail: {}",
+        errors.iter().join("\n")
+    )))
+}
+
+/// A set of variable names.
+/// Must be sorted by the variable name.
+type BooleanVariables = Vec<Name>;
+
+/// Generate sets of hypotheses to case-split over that are applicable to the target `defs`.
+/// - Construct hypotheses based on the variables used in the Boolean conditions in `defs`.
+/// - Excludes the literals in the `assumption` since it's already assumed to be true.
+/// - If there are variants with no extra Boolean variables, it will generate a no-hypothesis
+///   group, which contains only one empty clause.
+fn extract_boolean_hypotheses(
+    assumption: &Clause,
+    defs: &PossibleDefinitionsPerTypeCondition,
+) -> Vec<Vec<Clause>> {
+    // Collect sets of variables that can be used to case-split over.
+    let mut variable_groups = IndexSet::default();
+    for variant in defs.conditional_variants() {
+        let Some(remaining_condition) = variant.boolean_clause().subtract(assumption) else {
+            // Skip unsatisfiable variants.
+            continue;
+        };
+        // Collect variables from the remaining condition.
+        // Invariant: Clauses are expected to be sorted by the variable name.
+        let vars: BooleanVariables = remaining_condition
+            .literals()
+            .iter()
+            .map(|lit| lit.variable())
+            .cloned()
```

Comment:
Does this need to be cloned? We should be able to store references in `variable_groups` since we only need a reference when we `generate_clauses`.

Reply:
Unfortunately, `vars` outlives `remaining_condition`.